decode_typgen.go

```diff
@@ -0,0 +1,42 @@
+package msgpack
+
+import (
+	"reflect"
+)
+
+func (d *Decoder) newValue(t reflect.Type) reflect.Value {
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	gen := d.typeGen[t]
+
+	if gen == nil {
+		if d.typeGen == nil {
+			d.typeGen = map[reflect.Type]*ptrGen{}
+		}
+		gen = &ptrGen{typ: t, cap: 4096}
+		d.typeGen[t] = gen
+	}
+
+	return gen.next()
+}
+
+type ptrGen struct {
+	raw reflect.Value
+	typ reflect.Type
+	idx int
+	cap int
+}
+
+func (p *ptrGen) next() (v reflect.Value) {
+	if p.idx == p.cap || !p.raw.IsValid() {
+		p.raw = reflect.MakeSlice(reflect.SliceOf(p.typ), p.cap, p.cap)
```

Comment:
Looks like in some cases we will needlessly allocate 4096 elements and use only 1 of them. Is that true?

Reply:
This is true, I pushed a change to that and made this and alloc limits optional.